### PR TITLE
Features/multi language lookup all

### PR DESCRIPTION
**Pull request by Sargatos**
Modified the plugin and the tests accordingly.
You can test it out in the environment _Compliance - Dev Jojo_.
Since we now register the steps on all entities, I "merged" the 2 parts of the plugin into one.

### CODE REVIEW

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+                var pluginContext = context.GetDefaultPluginContext();
+
+                var multiLanguageEntity = GetMockedMultiLanguageEntity();
+                var expectedName = multiLanguageEntity.opc_namefrench;
+
+                var outputs = new ParameterCollection { { "BusinessEntity", multiLanguageEntity } };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "Retrieve";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1036);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+                var pluginContext = context.GetDefaultPluginContext();
+
+                var multiLanguageEntity = GetMockedMultiLanguageEntity();
+                var expectedName = multiLanguageEntity.opc_nameenglish;
+
+                var outputs = new ParameterCollection { { "BusinessEntity", multiLanguageEntity } };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "Retrieve";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1033);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntityCollection = GetMultiLanguageEntityCollection();
+
+                var outputs = new ParameterCollection { {"BusinessEntityCollection", multiLanguageEntityCollection} };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "RetrieveMultiple";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1036);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntityCollection = GetMultiLanguageEntityCollection();
+
+                var outputs = new ParameterCollection { { "BusinessEntityCollection", multiLanguageEntityCollection } };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "RetrieveMultiple";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1033);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntity = GetMockedMultiLanguageEntity();
+                var outputs = new ParameterCollection { { "BusinessEntity", multiLanguageEntity } };
+                var expectedName = "Technologie";
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "Retrieve";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1036);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntity = GetMockedMultiLanguageEntity();
+                var outputs = new ParameterCollection { { "BusinessEntity", multiLanguageEntity } };
+                var expectedName = "Technology";
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "Retrieve";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1033);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+            {
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntityCollection = GetMultiLanguageEntityCollection();
+                var outputs = new ParameterCollection { { "BusinessEntityCollection", multiLanguageEntityCollection } };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "RetrieveMultiple";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1036);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+            {
+                // Arrange
+                var context = new XrmFakedContext();
+                var pluginContext = context.GetDefaultPluginContext();
+                var multiLanguageEntityCollection = GetMultiLanguageEntityCollection();
+                var outputs = new ParameterCollection { { "BusinessEntityCollection", multiLanguageEntityCollection } };
+
+                pluginContext.OutputParameters = outputs;
+                pluginContext.MessageName = "RetrieveMultiple";
+                pluginContext.SharedVariables.Add("UserLocaleId", 1033);
```
Comment (joshcampbell191) [inline src/Compliance.Plugins.Tests/MultiLanguagePluginTests.cs]: Add constants for the locales.

File: src/Compliance.Plugins/MultiLanguagePlugin.cs [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+using System;
+using System.Linq;
+
+namespace Compliance.Plugins
+{
+    public partial class MultiLanguagePlugin : PluginBase
+    {
+        private readonly string preImageAlias = "PreImage";
+        private readonly string[] languages = new string[] { "english", "french" }; // Languages Supported
+        private readonly int[] locales = new int[] { 1033, 1036 }; // LCIDs of each language in the languages array
```
[the remark this answers is not in the document]
Comment (pierluc-ssoucy) [reply src/Compliance.Plugins/MultiLanguagePlugin.cs]: KVPs/Tuples with enums could work as well.

File: src/Compliance.Plugins/MultiLanguagePlugin.cs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            if (businessEntity.Attributes.ContainsKey("opc_name") && businessEntity["opc_name"].ToString().Contains(prefix))
+                businessEntity["opc_name"] = UnpackName(localContext, businessEntity.GetAttributeValue<string>("opc_name"));
+
+            foreach (var attribute in businessEntity.Attributes)
+            {
+                if (attribute.Key.EndsWith("id") && businessEntity[attribute.Key] is EntityReference entityReference && entityReference.Name != null && entityReference.Name.Contains(prefix))
+                {
+                    entityReference.Name = UnpackName(localContext, businessEntity.GetAttributeValue<EntityReference>(attribute.Key).Name);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (pierluc-ssoucy) [reply src/Compliance.Plugins/MultiLanguagePlugin.cs]: True. You can save the if with .OfType + .Where(a => a.Key.EndsWith("id") && a?.Name.Contains(prefix) ?? false)

**Pull request merged**